audio/pwaudio.c

```diff
@@ -0,0 +1,776 @@
+	uint32_t n_frames, req, index, n_bytes;
+	int32_t avail;
+
+	/*obtain a buffer to read from */
+	if ((b = pw_stream_dequeue_buffer(v->stream)) == NULL) {
+		pw_log_warn("out of buffers: %m");
+		return;
+	}
+
+	buf = b->buffer;
+	if ((p = buf->datas[0].data) == NULL)
+		return;
+	if ((req = b->requested * v->frame_size) == 0)
+		req = 4096 * v->frame_size;
+	n_frames = SPA_MIN(req, buf->datas[0].maxsize);
+	n_bytes = n_frames * v->frame_size;
+
+	/*get no of available bytes to read data from buffer */
+
+	avail = spa_ringbuffer_get_read_index(&v->ring, &index);
+
+	if (!v->enabled) {
+		avail = 0;
+	}
+
+	if (avail == 0) {
+		memset(p, 0, n_bytes);
+	} else {
+		if (avail < (int32_t) n_bytes)
+			n_bytes = avail;
+
+		spa_ringbuffer_read_data(&v->ring,
+					 v->buffer, RINGBUFFER_SIZE,
+					 index & RINGBUFFER_MASK, p, n_bytes);
+
+		index += n_bytes;
+		spa_ringbuffer_read_update(&v->ring, index);
+	}
+
+	buf->datas[0].chunk->offset = 0;
+	buf->datas[0].chunk->stride = v->frame_size;
+	buf->datas[0].chunk->size = n_bytes;
+
+	/*queue the buffer for playback */
+	pw_stream_queue_buffer(v->stream, b);
+}
+
+/*output data processing function to generate stuffs in the buffer*/
+static void capture_on_process(void *data)
+{
+	PWVoice *v = (PWVoice *) data;
+	void *p;
+	struct pw_buffer *b;
+	struct spa_buffer *buf;
+	int32_t filled;
+	uint32_t index, offs, n_bytes;
+
+	/*obtain a buffer */
+	if ((b = pw_stream_dequeue_buffer(v->stream)) == NULL) {
+		pw_log_warn("out of buffers: %m");
+		return;
+	}
+
+	/*Write data into buffer */
+	buf = b->buffer;
+	if ((p = buf->datas[0].data) == NULL)
+		return;
+	offs = SPA_MIN(buf->datas[0].chunk->offset, buf->datas[0].maxsize);
+	n_bytes =
+	    SPA_MIN(buf->datas[0].chunk->size, buf->datas[0].maxsize - offs);
+
+	filled = spa_ringbuffer_get_write_index(&v->ring, &index);
+
+	if (!v->enabled) {
+		n_bytes = 0;
+	}
+
+	if (filled < 0) {
+		pw_log_warn("%p: underrun write:%u filled:%d",
+			    p, index, filled);
+	} else if ((uint32_t) filled + n_bytes > RINGBUFFER_SIZE) {
+		pw_log_warn("%p: overrun write:%u filled:%d + size:%u > max:%u",
+			    p, index, filled, n_bytes, RINGBUFFER_SIZE);
+	}
+	spa_ringbuffer_write_data(&v->ring,
+				  v->buffer, RINGBUFFER_SIZE,
+				  index & RINGBUFFER_MASK,
+				  SPA_PTROFF(p, offs, void), n_bytes);
+	index += n_bytes;
+	spa_ringbuffer_write_update(&v->ring, index);
+
+	/*queue the buffer for playback */
+	pw_stream_queue_buffer(v->stream, b);
+}
+
+static void on_stream_state_changed(void *_data, enum pw_stream_state old,
+				    enum pw_stream_state state,
+				    const char *error)
+{
+	PWVoice *v = (PWVoice *) _data;
+
+	printf("stream state: \"%s\"\n", pw_stream_state_as_string(state));
+
+	switch (state) {
+	case PW_STREAM_STATE_ERROR:
+	case PW_STREAM_STATE_UNCONNECTED:
+		{
+			break;
+		}
+	case PW_STREAM_STATE_PAUSED:
+		printf("node id: %d\n", pw_stream_get_node_id(v->stream));
+		break;
+	case PW_STREAM_STATE_STREAMING:
+	default:
```

Comment:
I do not see the definition of the enum, but if all states are listed here, you could skip the default case, it will never happen (the compiler will check that all states are handled for you).

Otherwise, this seems as you would only need to handle `case PW_STREAM_STATE_PAUSED:` and `default:`.

Reply:
actually not all states were listed here, I think it's better to just list all states and do away with the default.